app/models/authenticator/database.rb

```diff
@@ -13,7 +13,15 @@ def uses_stored_password?
     def _authenticate(username, password, _request)
       user = case_insensitive_find_by_userid(username)
 
-      user && user.authenticate_bcrypt(password)
+      return [false, _('Your account has been locked due to too many failed login attempts, please contact the administrator.')] if user&.locked?
+
+      if user&.authenticate_bcrypt(password) # Authenticate if the username matches
+        user.unlock! # Reset the number of failed logins
+        return true
+      end
+
+      user&.fail_login! # Increase the number of failed login attempts
+      [false, _("Sorry, the username or password you entered is incorrect.")]
```

Comment:
I don't think we should put the "Sorry" bit in there.  Prefer `The username or password you entered is incorrect.`.